source/Halibut/ServiceModel/PendingRequestQueueAsync.cs

```diff
@@ -0,0 +1,250 @@
+                    {
+                        if (!transferBegun)
+                        {
+                            completed = true;
+                            log.Write(EventType.MessageExchange, "Request {0} was cancelled before it could be collected by the polling endpoint", request);
+                            throw;
+                        }
+
+                        log.Write(EventType.MessageExchange, "Request {0} was cancelled after it had been collected by the polling endpoint and will not be cancelled", request);
+                    }
+                }
+
+                var waitForTransferToComplete = false;
+                using (await transferLock.LockAsync())
+                {
+                    if (transferBegun)
+                    {
+                        waitForTransferToComplete = true;
+                    }
+                    else
+                    {
+                        completed = true;
+                    }
+                }
+
+                if (waitForTransferToComplete)
+                {
+                    responseSet = await WaitForResponseToBeSet(request.Destination.PollingRequestMaximumMessageProcessingTimeout, CancellationToken.None);
+                    if (responseSet)
+                    {
+                        // We end up here when the request is cancelled but already being transferred so we need to adjust the log message accordingly
+                        if (cancelled)
+                        {
+                            log.Write(EventType.MessageExchange, "Request {0} was collected by the polling endpoint", request);
+                        }
+                        else
+                        {
+                            log.Write(EventType.MessageExchange, "Request {0} was eventually collected by the polling endpoint", request);
+                        }
+                    }
+                    else
+                    {
+                        SetResponse(ResponseMessage.FromException(request, new TimeoutException($"A request was sent to a polling endpoint, the polling endpoint collected it but did not respond in the allowed time ({request.Destination.PollingRequestMaximumMessageProcessingTimeout}), so the request timed out.")));
+                    }
+                }
+                else
+                {
+                    log.Write(EventType.MessageExchange, "Request {0} timed out before it could be collected by the polling endpoint", request);
+                    SetResponse(ResponseMessage.FromException(request, new TimeoutException($"A request was sent to a polling endpoint, but the polling endpoint did not collect the request within the allowed time ({request.Destination.PollingRequestQueueTimeout}), so the request timed out.")));
+                }
+            }
+
+            async Task<bool> WaitForResponseToBeSet(TimeSpan timeout, CancellationToken cancellationToken)
```

Comment:
This used to be a simple call to `waiter.Wait(request.Destination.PollingRequestQueueTimeout, cancellationToken)`.

But the new libraries do not support a timeout.

This code works, but I am open to better ways of doing this.

Reply:
Possible using a `WaitHandle` with `WaitOne` could avoid this. However I am not sure which direction I trust more.

In any case this is sufficient.